src/hotspot/share/gc/shenandoah/shenandoahReferenceProcessor.cpp

```diff
@@ -544,10 +565,29 @@ void ShenandoahReferenceProcessor::process_references(ShenandoahPhaseTimings::Ph
 void ShenandoahReferenceProcessor::enqueue_references_locked() {
   // Prepend internal pending list to external pending list
   shenandoah_assert_not_in_cset_except(&_pending_list, _pending_list, ShenandoahHeap::heap()->cancelled_gc() || !ShenandoahLoadRefBarrier);
+
+  // During reference processing, we maintain a local list of references that are identified by
+  //   _pending_list and _pending_list_tail.  _pending_list_tail points to the next field of the last Reference object on
+  //   the local list.
+  //
+  // There is also a global list of reference identified by Universe::_reference_pending_list
+
+  // The following code has the effect of:
+  //  1. Making the global Universe::_reference_pending_list point to my local list
+  //  2. Overwriting the next field of the last Reference on my local list to point at the previous head of the
+  //     global Universe::_reference_pending_list
+
+  oop former_head_of_global_list = Universe::swap_reference_pending_list(_pending_list);
   if (UseCompressedOops) {
-    *reinterpret_cast<narrowOop*>(_pending_list_tail) = CompressedOops::encode(Universe::swap_reference_pending_list(_pending_list));
+    *reinterpret_cast<narrowOop*>(_pending_list_tail) = CompressedOops::encode(former_head_of_global_list);
   } else {
-    *reinterpret_cast<oop*>(_pending_list_tail) = Universe::swap_reference_pending_list(_pending_list);
+    *reinterpret_cast<oop*>(_pending_list_tail) = former_head_of_global_list;
+  }
+  ShenandoahHeap* heap = ShenandoahHeap::heap();
+  if (ShenandoahCardBarrier) {
+    if (heap->is_in_old(_pending_list_tail) && heap->is_in_young(former_head_of_global_list)) {
+      heap->mark_card_as_dirty(_pending_list_tail);
+    }
```

Comment:
Can we just replace this section with `set_oop_field(_pending_list_tail, former_head_of_global_list)` ?

Reply:
Have made this improvement, and am retesting.